### DISCUSSION TITLE
Authentication Keycloak + Cloudflare

### DISCUSSION THREAD

**Discussion by FluxSon** [outline/outline#5617] [Self Hosting] [answered]
Hi guys, 
I've been breaking my head over this for the past couple of hours and I can't seem to find what's wrong with my set-up.
I have Keycloak running through a cloudflare tunnel and on Keycloak's side everything seems to work fine.
Similar Set-up for Outline, auth set-up followed by a mixture of these 2 guides: https://medium.com/@ange.cesari/how-to-deploy-outline-wiki-fully-self-hosted-no-tls-with-docker-compose-e50b57baaf73 & https://blog.gurucomputing.com.au/Knowledgebases%20with%20Outline/Installing%20Outline%20Knowledgebase/#configuring-oidc

My Login after it goes through the Keycloak login page and back to outline, shows the following:
![image](https://github.com/outline/outline/assets/23508501/b89d281a-ce09-4d43-bcfa-bd656eae9173)

On Outline docker logs I'm getting the following error when trying to login:

> 2023-07-27T11:42:43.895014108Z ERR Error during authentication | error=invalid json response body at https://my-domain.com/realms/master/protocol/openid-connect/userinfo reason: Unexpected token < in JSON at position 0 stack=FetchError: invalid json response body at https://my-domain.com/realms/master/protocol/openid-connect/userinfo reason: Unexpected token < in JSON at position 0
    at /opt/outline/node_modules/node-fetch/lib/index.js:273:32
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async _passportOauth.Strategy.userProfile (/opt/outline/build/plugins/oidc/server/auth/oidc.js:23:22) 

Although in Keycloak I am seeing it as a succesful login:
![image](https://github.com/outline/outline/assets/23508501/e20d26ef-6605-4068-b16c-4ec8ca0eb96a)

Part of my Docker Compose for outline:
>  outline:
    image: outlinewiki/outline:latest
    user: root
    restart: always
    container_name: outline
    command: sh -c "yarn start --env=production-ssl-disabled"
    depends_on:
      - outline_postgres
      - outline_redis
      - outline_minio
    environment:
      - PGSSLMODE=disable
      - SECRET_KEY=${SECRET_KEY}
      - UTILS_SECRET=${UTILS_SECRET}
      - DATABASE_URL=postgres://outline:${POSTGRES_PASSWORD}@outline_postgres:5432/outline
      - REDIS_URL=redis://outline_redis:6379
      - URL=${WIKI_URL}
      - FORCE_HTTPS=false
      - AWS_ACCESS_KEY_ID=minio
      - AWS_REGION=us-east-1
      - AWS_SECRET_ACCESS_KEY=${MINIO_ROOT_PASSWORD}
      - AWS_S3_UPLOAD_BUCKET_URL=${WIKIDATA_URL}
      - AWS_S3_UPLOAD_BUCKET_NAME=outline
      - AWS_S3_FORCE_PATH_STYLE=true
      - AWS_S3_ACL=private
      - SMTP_HOST=${SMTP_HOST}
      - SMTP_PORT=${SMTP_PORT}
      - SMTP_FROM_EMAIL=${SMTP_FROM_EMAIL}
      - SMTP_REPLY_EMAIL=${SMTP_REPLY_EMAIL}
      - SMTP_SECURE=${SMTP_SECURE}
      - OIDC_CLIENT_ID=${OIDC_CLIENT_ID}
      - OIDC_CLIENT_SECRET=${OIDC_CLIENT_SECRET}
      - OIDC_AUTH_URI=${OIDC_AUTH_URI}
      - OIDC_TOKEN_URI=${OIDC_TOKEN_URI}
      - OIDC_USERINFO_URI=${OIDC_USERINFO_URI}
      - OIDC_DISPLAY_NAME=${OIDC_DISPLAY_NAME}
      - OIDC_USERNAME_CLAIM=email
      - OIDC_SCOPES=openid profile email

Would greatly appreciate some help for pointing me in the right direction.
Kind regards

**Comment by tommoor** [comment]
> Although in Keycloak I am seeing it as a succesful login

Okay, that's good – it seems like the OAuth process completes successfully. After the auth we hit the user info endpoint to get the profile details and that's what is failing. Are you sure it's correct and enabled? Non-json response would generally mean incorrect endpoint or some sort of server error.

**Comment by stereo9** [comment]
could you share your client settings?

**Reply by stereo9** [reply]
mhh i remember having similar issues during setup but i don't remember how i fixed it. i have the same settings in keycloak.. 

can you check if the environment variables are set incorrectly? did you try hardcoding them?

this is how it looks in my case:

      - OIDC_CLIENT_ID=outline
      - OIDC_CLIENT_SECRET=***************
      - OIDC_AUTH_URI=https://**************.com/auth/realms/**************/protocol/openid-connect/auth
      - OIDC_TOKEN_URI=https://**************.com/auth/realms/**************/protocol/openid-connect/token
      - OIDC_USERINFO_URI=https://**************.com/auth/realms/**************/protocol/openid-connect/userinfo
      - OIDC_USERNAME_CLAIM=email
      - OIDC_DISPLAY_NAME=Keycloak
      - OIDC_SCOPES=openid profile email
      

my keycloak event logs look like this:
![image](https://github.com/outline/outline/assets/55000172/86c19f57-0015-4a6e-88ae-a5aab65e8225)


i'm more the neighbor's uncle kind of guy who offers to help fix the broken down peugeot 206 rather than being an expert but i can imagine this could be related to using https as well.

**Reply by FluxSon** [reply]
Hmm, I did try to hardcode the settings but alas it didn't seem to have any effect. 
I tried creating a new realm in keycloak, replies in postman look solid even when using the same URLs as in the outline config. 

Nothing seems to work here 😢

**Comment by FluxSon** [comment, marked as answer]
I've found the issue!
It was cloudflare, the 2nd request was filtered out by the Firewall rules 🗡️ 
Thank you for the possible problems and suggestions for solutions!!
